plugins/ftdidmx/FtdiWidget-libftdi.cpp

```diff
@@ -47,36 +54,169 @@ using std::vector;
+      ftdi_device_list* current_device = list;
+
+      while (current_device != NULL) {
+        struct usb_device *dev = current_device->dev;
+        current_device = current_device->next;
+        i++;
+
+        if (!dev) {
+          OLA_WARN << "Device returned from ftdi_usb_find_all was NULL";
+          continue;
+        }
+
+        char serial[256];
+        char name[256];
+        char vendor[256];
+
+        int r = ftdi_usb_get_strings(ftdi, dev,
+                                    vendor, sizeof(vendor),
+                                    name, sizeof(name),
+                                    serial, sizeof(serial));
+
+        // libftdi doesn't enumerate error codes,
+        // -9 is 'get serial number failed', not ideal but workable.
+        if (r < 0 && r != -9) {
+          OLA_WARN << "Unable to fetch string information from USB device: "
+                   << ftdi_get_error_string(ftdi);
+          continue;
+        }
+
+        string v = string(vendor);
+        string sname = string(name);
+        string sserial = string(serial);
+        if (sserial == "?" || r == -9) {
+          // this means there wasn't a serial number
+          sserial.clear();
+        }
+        OLA_INFO << "Found FTDI device. Vendor: '" << v << "', Name: '"
+                 << sname << "', Serial: '" << sserial << "'";
+        ToUpper(&v);
+        if (std::string::npos != v.find("FTDI") ||
+            std::string::npos != v.find("KMTRONIC") ||
+            std::string::npos != v.find("WWW.SOH.CZ")) {
+          widgets->push_back(FtdiWidgetInfo(sname,
+                                            sserial,
+                                            i,
+                                            vid,
+                                            pids[cur_pid]));
+        } else {
+          OLA_INFO << "Unknown FTDI device with vendor string: '" << v << "'";
+        }
+      }  // while (list != NULL)
+      OLA_DEBUG << "Freeing list";
+      ftdi_list_free(&list);
+    }
+  }
+  ftdi_free(ftdi);
+}
+
+FtdiInterface::FtdiInterface(const FtdiWidget * parent,
+                             const ftdi_interface interface)
+      : m_parent(parent),
+        m_interface(interface) {
+  memset(&m_handle, '\0', sizeof(struct ftdi_context));
   ftdi_init(&m_handle);
 }
 
-FtdiWidget::~FtdiWidget() {
-  if (IsOpen())
+
+FtdiInterface::~FtdiInterface() {
+  if (IsOpen()) {
     Close();
+  }
   ftdi_deinit(&m_handle);
 }
 
+bool FtdiInterface::SetInterface() {
+  OLA_INFO << "Setting interface to: " << m_interface;
+  if (ftdi_set_interface(&m_handle, m_interface) < 0) {
+    OLA_WARN << m_parent->Name() << " " << ftdi_get_error_string(&m_handle);
+    return false;
+  } else {
+    return true;
+  }
+}
 
-bool FtdiWidget::Open() {
-  if (Serial().empty()) {
-    OLA_WARN << Name() << " has no serial number, "
+bool FtdiInterface::Open() {
+  if (m_parent->Serial().empty()) {
+    OLA_WARN << m_parent->Name() << " has no serial number, "
```

Comment:
Within the usbdmx plugin, we fail to add the second device if there's already one with no serial present, its probably worth doing the same here.